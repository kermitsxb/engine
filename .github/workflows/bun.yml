# This workflow will do a clean installation of node dependencies, cache/restore them, build the source code and run tests across different versions of node
# For more information see: https://docs.github.com/en/actions/automating-builds-and-tests/building-and-testing-nodejs

name: Bun CI

on:
  push:
    branches: [ "main" ]
  pull_request:
    branches: [ "main" ]

jobs:
  build:
    runs-on: ubuntu-latest
    
    strategy:
      matrix:
        bun-version: [1.x]

    steps:
    - uses: actions/checkout@v3
    - uses: oven-sh/setup-bun@v1
      with:
        bun-version: latest
    - run: bun install
    - run: bun run lint
    - run: bun run build --if-present
    - run: bunx playwright install
    - run: bunx @puppeteer/browsers install chrome@stable --path $HOME/.cache/puppeteer
    - run: bun run test
    #- name: Release
      #env:
        #GITHUB_TOKEN: ${{ secrets.GH_TOKEN }}
        #NPM_TOKEN: ${{ secrets.NPM_TOKEN }}
      #run: bunx semantic-release
    
