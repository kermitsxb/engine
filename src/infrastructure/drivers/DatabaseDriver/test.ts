import type { IDatabaseTableDriver } from '@adapter/spi/drivers/DatabaseTableSpi'
import BunTest, { afterAll } from 'bun:test'

export function testDatabaseTableDriver(
  { describe, beforeAll, it, expect }: typeof BunTest,
  setup: (date: Date) => Promise<IDatabaseTableDriver>,
  teardown?: () => Promise<void>
) {
  let databaseTable: IDatabaseTableDriver
  const date = new Date()

  beforeAll(async () => {
    databaseTable = await setup(date)
  })

  afterAll(async () => {
    if (teardown) await teardown()
  })

  describe('read', () => {
    it('should return a row with a filter', async () => {
      // WHEN
      const row = await databaseTable.read({
        field: 'name',
        operator: 'Is',
        value: 'John Doe',
      })

      // THEN
      expect(row).toStrictEqual({
        id: '1',
        fields: { name: 'John Doe' },
        created_at: date,
        updated_at: date,
      })
    })

    it('should not return a row with an empty "or" filter', async () => {
      // WHEN
      const row = await databaseTable.read({ or: [] })

      // THEN
      expect(row).toBeUndefined()
    })

    it('should not return a row with an empty "and" filter', async () => {
      // WHEN
      const row = await databaseTable.read({ and: [] })

      // THEN
      expect(row).toBeUndefined()
    })

    it('should not return a row with a filter', async () => {
      // WHEN
      const row = await databaseTable.read({
        field: 'name',
        operator: 'Is',
        value: 'Jane Doe',
      })

      // THEN
      expect(row).toBeUndefined()
    })

    it('should not return a row', async () => {
      // WHEN
      const row = await databaseTable.read({
        field: 'name',
        operator: 'Is',
        value: 'Jane Doe',
      })

      // THEN
      expect(row).toBeUndefined()
    })
  })

  describe('list', () => {
    it('should return a list of rows with no filter', async () => {
      // WHEN
      const rows = await databaseTable.list()

      // THEN
      expect(rows).toStrictEqual([
        {
          id: '1',
          fields: { name: 'John Doe' },
          created_at: date,
          updated_at: date,
        },
      ])
    })

    it('should return a list of rows with an empty "or" filter', async () => {
      // WHEN
      const rows = await databaseTable.list({ or: [] })

      // THEN
      expect(rows).toHaveLength(1)
    })

    it('should return a list of rows with an empty "and" filter', async () => {
      // WHEN
      const rows = await databaseTable.list({ and: [] })

      // THEN
      expect(rows).toHaveLength(1)
    })

    it('should return an empty list of rows with a filter', async () => {
      // WHEN
      const rows = await databaseTable.list({
        field: 'name',
        operator: 'Is',
        value: 'Jane Doe',
      })

      // THEN
      expect(rows).toHaveLength(0)
    })
  })
}
