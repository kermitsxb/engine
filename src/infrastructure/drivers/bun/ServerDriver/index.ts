import type { IServerDriver } from '/adapter/spi/drivers/ServerSpi'
import type { DeleteDto, GetDto, PatchDto, PostDto, RequestDto } from '/adapter/spi/dtos/RequestDto'
import type { ServerConfig } from '/domain/services/Server'
import type { Response } from '/domain/entities/Response'
import { ElysiaDriver } from './ElysiaDriver'

export class ServerDriver implements IServerDriver {
  private _server: ElysiaDriver

  constructor(config: ServerConfig) {
    this._server = new ElysiaDriver(config)
  }

  get = async (path: string, handler: (getDto: GetDto) => Promise<Response>) => {
    await this._server.get(path, handler)
  }

  post = async (path: string, handler: (postDto: PostDto) => Promise<Response>) => {
    await this._server.post(path, handler)
  }

  patch = async (path: string, handler: (patchDto: PatchDto) => Promise<Response>) => {
    await this._server.patch(path, handler)
  }

  delete = async (path: string, handler: (deleteDto: DeleteDto) => Promise<Response>) => {
    await this._server.delete(path, handler)
  }

  notFound = async (handler: (requestDto: RequestDto) => Promise<Response>) => {
    await this._server.notFound(handler)
  }

  start = async (): Promise<number> => {
    return this._server.start()
  }

  stop = async () => {
    await this._server.stop()
  }
}
