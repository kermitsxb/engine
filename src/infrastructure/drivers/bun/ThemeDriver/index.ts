import type { IThemeDriver } from '/adapter/spi/drivers/ThemeSpi'
import type { ThemeConfig } from '/domain/services/Theme'
import { TailwindCSSDriver } from './TailwindCSSDriver'

export class ThemeDriver implements IThemeDriver {
  private _driver: IThemeDriver

  constructor(config: ThemeConfig) {
    const { type } = config
    switch (type) {
      case 'tailwindcss':
        this._driver = new TailwindCSSDriver(config)
        break
      default:
        throw new Error(`Unsupported Theme driver: ${type}`)
    }
  }

  buildCss = async (): Promise<string> => {
    return this._driver.buildCss()
  }

  buildJs = async (): Promise<string> => {
    return this._driver.buildJs()
  }
}
