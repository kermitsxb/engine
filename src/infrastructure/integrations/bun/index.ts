import type { Integrations } from '@adapter/spi/integrations'
import type { NotionConfig } from '@domain/integrations/Notion'
import type { PappersConfig } from '@domain/integrations/Pappers'
import type { QontoConfig } from '@domain/integrations/Qonto'
import type { NgrokConfig } from '@domain/integrations/Ngrok'

import { NotionIntegration } from '../NotionIntegration'
import { PappersBunIntegration } from './PappersBunIntegration'
import { QontoBunIntegration } from './QontoBunIntegration'
import { NgrokBunIntegration } from './NgrokBunIntegration'

export const integrations: Integrations = {
  notion: (config?: NotionConfig) => new NotionIntegration(config),
  pappers: (config?: PappersConfig) => new PappersBunIntegration(config),
  qonto: (config?: QontoConfig) => new QontoBunIntegration(config),
  ngrok: (config?: NgrokConfig) => new NgrokBunIntegration(config),
}
