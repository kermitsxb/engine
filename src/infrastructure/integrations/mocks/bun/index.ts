import type { Integrations } from '@adapter/spi/integrations'
import type { NotionConfig } from '@domain/integrations/Notion'
import type { PappersConfig } from '@domain/integrations/Pappers'
import type { QontoConfig } from '@domain/integrations/Qonto'
import type { NgrokConfig } from '@domain/integrations/Ngrok'

import { NotionBunIntegration } from './NotionIntegration'
import { PappersBunIntegration } from './PappersIntegration'
import { QontoBunIntegration } from './QontoIntegration'
import { NgrokBunIntegration } from './NgrokIntegration'

export const mocks: Integrations = {
  notion: (config?: NotionConfig) => new NotionBunIntegration(config),
  pappers: (config?: PappersConfig) => new PappersBunIntegration(config),
  qonto: (config?: QontoConfig) => new QontoBunIntegration(config),
  ngrok: (config?: NgrokConfig) => new NgrokBunIntegration(config),
}
