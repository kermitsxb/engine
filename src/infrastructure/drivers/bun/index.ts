import type { Drivers } from '/adapter/spi/drivers'
import type { DatabaseConfig } from '/domain/services/Database'
import type { MonitorsConfig } from '/domain/services/Monitor'
import type { ServerConfig } from '/domain/services/Server'
import type { ThemeConfig } from '/domain/services/Theme'

import { DatabaseDriver } from './DatabaseDriver'
import { MonitorDriver } from './MonitorDriver'
import { ServerDriver } from './ServerDriver'
import { ThemeDriver } from './ThemeDriver'

import { drivers as common } from '../common'

export const drivers: Drivers = {
  ...common,
  monitor: (config: MonitorsConfig) => new MonitorDriver(config),
  database: (config: DatabaseConfig) => new DatabaseDriver(config),
  server: (config: ServerConfig) => new ServerDriver(config),
  theme: (config: ThemeConfig) => new ThemeDriver(config),
}
