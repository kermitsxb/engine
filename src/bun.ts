import { drivers } from '/infrastructure/drivers/bun'
import { integrations } from '/infrastructure/integrations/bun'
import App from '/adapter/api'
import type { Drivers } from '/adapter/spi/drivers'
import type { Integrations } from '/adapter/spi/integrations'

export { mocks } from '/infrastructure/integrations/bun/mocks'
export { NotionTablePage } from '/domain/integrations/Notion/NotionTablePage'
export { Record as DatabaseTableRecord } from '/domain/entities/Record'

export default class extends App {
  constructor(options?: { drivers?: Partial<Drivers>; integrations?: Partial<Integrations> }) {
    const customDrivers = options?.drivers ?? {}
    const customIntegrations = options?.integrations ?? {}
    super({ ...drivers, ...customDrivers }, { ...integrations, ...customIntegrations })
  }
}
