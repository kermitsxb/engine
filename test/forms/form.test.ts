import Tester, { expect, describe, it } from 'bun:test'
import { Mock, type Config } from '/test/bun'

const mock = new Mock(Tester, { drivers: ['Database'] })

mock.page(({ app, browser, drivers }) => {
  describe('on GET', () => {
    const config: Config = {
      name: 'App',
      version: '1.0.1',
      forms: [
        {
          name: 'user',
          path: 'user',
          title: 'Form title',
          description: 'Form description',
          table: 'users',
          inputs: [
            {
              field: 'name',
              label: 'Name',
              required: true,
            },
          ],
          submitLabel: 'Save',
        },
      ],
      tables: [
        {
          name: 'users',
          fields: [
            {
              name: 'name',
              type: 'SingleLineText',
            },
          ],
        },
      ],
    }

    it('should display the form title', async () => {
      // GIVEN
      const { url } = await app.start(config)

      // WHEN
      await browser.page.goto(`${url}/forms/user`)

      // THEN
      expect(browser.page.title()).resolves.toBe('Form title')
      expect(browser.page.content()).resolves.toContain('Form title')
    })

    it('should display the form description', async () => {
      // GIVEN
      const { url } = await app.start(config)

      // WHEN
      await browser.page.goto(`${url}/forms/user`)

      // THEN
      expect(browser.page.content()).resolves.toContain('Form description')
    })

    it('should display the form inputs', async () => {
      // GIVEN
      const { url } = await app.start(config)

      // WHEN
      await browser.page.goto(`${url}/forms/user`)

      // THEN
      expect(browser.page.content()).resolves.toContain('Name')
    })

    it('should display the form submit button', async () => {
      // GIVEN
      const { url } = await app.start(config)

      // WHEN
      await browser.page.goto(`${url}/forms/user`)

      // THEN
      expect(browser.page.content()).resolves.toContain('Save')
    })

    it('should create a record when the form is submitted', async () => {
      // GIVEN
      const { page } = browser
      const table = drivers.database.table(config.tables![0])
      const { url } = await app.start(config)

      // WHEN
      await page.goto(`${url}/forms/user`)
      await page.type('input[name="name"]', 'John Doe')
      await page.click('button[type="submit"]')
      await page.waitForNavigation()

      // THEN
      const records = await table.list()
      expect(records).toHaveLength(1)
      expect(records[0].fields.name).toBe('John Doe')
    })
  })
})
