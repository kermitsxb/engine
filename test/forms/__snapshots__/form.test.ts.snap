// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`ui on GET should match the form snapshot 1`] = `"<html lang="en"><head><title>Form title</title><meta name="description" content="Form description"></head><body><div><h1>Form title</h1><p>Form description</p><form action="/api/table/users" method="POST"><div><label for="name">Name</label><input type="text" required="" name="name"></div><button type="submit">Save</button></form></div></body></html>"`;
